### PR TITLE
feat: add hover language feature for modules and module function calls

### LINKED ISSUE #48: hover docs

### PR DESCRIPTION
**Pull request by sineed**
closes #48

https://github.com/elixir-tools/next-ls/assets/929818/780eac52-a684-4106-9728-1ed6f7328521

It figures out if module or function call is hovered using `Code.Fragment.surround_context/3` and gets docs via `Code.fetch_docs/1` functions. It also supports module aliases in three forms:

What is supported:
- [x] the module from this file
- [x] aliased modules and their functions:
    - [x] Basic aliasing
    - [x] Aliasing with `:as` option
    - [x] Multi alias
- [x] `Kernel` and `Kernel.SpecialForms` functions
- [x] Erlang modules and functions
- [x] Structs
- [x] Imported module functions

### CODE REVIEW

**Comment by sineed** [comment]
The missing piece is imported functions. I didn't cover it as it is connected to compile tracer and I know that the `SymbolTable` is a subject of change. I think better to have a follow-up PR for imports.

File: lib/next_ls/hover.ex [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+      other ->
+        other
+    end
+  end
+
+  defp send_fetch_docs_request(lsp, module) do
+    Runtime.call(lsp.assigns.runtime, {Code, :fetch_docs, [module]})
+  end
+
+  defp find_aliased_module(document, module) do
```
[the remark this answers is not in the document]
Comment (mhanberg) [reply lib/next_ls/hover.ex:92]: My intuition for how to implement this feature was going to be using the reference table to find the right thing and then look up the docs (or, just stick the docs in the symbol table), but i did have concerns about being able to lookup docs for code you've written but hasn't compiled yet, which using your implementation would solve for.

I'll probably get time tonight to review this, but even if we want to use the ref and symbol tables, we can probably merge this and change it later.

[the remark this answers is not in the document]
Comment (sineed) [reply lib/next_ls/hover.ex:92]: Sounds a better idea. My approach don't support imported functions which you can find quite often in files using `Phoenix.LiveView` and `Phoenix.Controller`. But the reference table has this info OOTB.
As for the code not yet compiled I don't see it a big problem as long as user needs to just save the file. If you don't agree there may be two steps:

1. It checks references table
2. If nothing found it uses `Code.Fragment.surround_context/2`

I'm moving it back to draft but waiting for your feedback

[the remark this answers is not in the document]
Comment (sineed) [reply lib/next_ls/hover.ex:92]: not quite satisfied with results: only 6 out of 15 assertions that I added for the hovering test were passed. I may need to extend code in Tracer a bit

**Comment by mhanberg** [comment]
@sineed I apologize for probably creating a horrid rebase situation, but I switched Next LS to use sqlite3 instead of dets.

I am still interested in merging this work, please let me know if you'll get the time to rebase this on main.

Thanks again!

**Comment by leonqadirie** [comment]
Apologies for hijacking; would we be open to use explicit language identifiers in the generated markdown code-fences, so editors like helix could highlight the hover syntax?

[Analogous to this issue opened in ElixirLS.](https://github.com/elixir-lsp/elixir-ls/issues/694)

**Comment by mhanberg** [comment]
> Apologies for hijacking; would we be open to use explicit language identifiers in the generated markdown code-fences, so editors like helix could highlight the hover syntax?
> 
> [Analogous to this issue opened in ElixirLS.](https://github.com/elixir-lsp/elixir-ls/issues/694)

Potentially, the problem would be potentially mis highlighting code snippets that are of js or html/heex but are lacking the name in the fence

but, i think exdoc assumes its elixir if you don't specify, so there is already a precedence of that.

This would be a nice enhancement once this lands, as well as translating the erlang documentation into markdown as well (it's stored in a pseduo html syntax tree)

**Comment by mhanberg** [comment]
Superceded by #220

**Pull request closed without merging**